Mute column

ericmehl opened the pull request:
This adds a new column to the Light Editor to allow muting of lights, turning them off completely, without changing their intensity.

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon reviewed (changes requested):
Thanks Eric!

In addition to the inline comments, I've got a few observations from testing it interactively.

- There were definitely times when I wanted to get access to the actual plugs behind the state, like you can for the other cells. The only way I found to achieve this was to right-click, which as well as showing the menu had the side-effect of selecting the cell. And then pressing Enter to pop up the editor. Or using the keyboard to perform selection and then pressing Enter. Without access to the plugs, I felt like I was flying blind, so I think we need to cater for this a bit more directly somehow.
- Muscle memory from the Spreadsheet led to me double-clicking the icon expecting a popup for the plug editor. This doesn't work though - it toggles twice instead. _Unless_ a different cell is selected at the time (a regular one like intensity), in which case an editor for that _other_ cell is popped up.
- I found the icons a bit confusing/misleading. Because the column is called "Mute", and we're editing the mute state, it feels like the brightest icon should be reserved for when the light is muted. I think maybe something based around an off switch or a strike-through might capture this a bit more?

Those observations are leading me to think we should possibly favour the existing spreadsheet-style interaction we use for the other LightEditor cells, over the more-immediate-but-with-less-access-to-the-node-graph single-click implemented here. I've asked @murraystevenson to take it for a test drive too to see if I'm just being a grumpy old man or not. There's definitely some tension because the upcoming HierarchyView inclusions/exclusions columns seem to be benefiting from the more immediate interaction (but then again, they're not related to graph edits). So I'm definitely open to any further thoughts you might have about compromises or other improvements to the interaction...

Cheers...
John

ericmehl commented:
Another issue with the interaction currently is that you can't toggle multiple mutes on and off at the same time. That seems like something that might be pretty common.

I see what you mean about access to the plugs. Among other things there's no easy way to disable a tweak if you want to allow the upstream value to pass through. The spreadsheet behavior seems like it would be an ok interaction pattern. I hadn't realized spreadsheets directly toggle a boolean without a plug popup.

So to match the spreadsheet behavior, we would :
1. Single click to select
2. Double click to toggle
3. Add an "Edit Plug" or similar item to the right click menu to give a popup

Does that sound right?

johnhaddon commented:
> So to match the spreadsheet behavior, we would...Does that sound right?

Yep, I think so. Seems like some (most?) of that might be doable on the InspectorColumn, so double-click applies to any Bool value, and "Edit..." applies to all properties?

ericmehl commented:
The latest push has what I think is better interaction. Pretty much what we discussed above and on our call yesterday - matching the spreadsheet and adding a right-click menu to edit the plug itself.

I also changed the icons so that the "mute" state ticked on is the more dominant visual. I'm liking not having any icon for lights that are inheriting their mute state from a parent location. For anyone who wants to play with it, the commented sections started at line 269 in `LightEditorBinding.cpp` will enable faded icons for locations that inherit their mute state. Perhaps it's better to be more explicit about the state so you don't have to visually track it from the parent, possibly multiple locations up?

There are a couple other icons in there too, that can be used by switching out the file names at line 322 of the same file.

johnhaddon commented:
Thanks Eric - here's my tentative tuppence worth in the mute/solo bikeshedding of 2022 :

- Having a "positive" icon for when the light is muted is a definite win. I'm not sure we've struck on the right icon yet though, and in particular I wonder if my "switch" idea was a bad one, because it implies the single-click-to-edit that you'd get in the NodeEditor. The red circle alone would be a bit abstract, but might be sufficient? Do you have any thoughts on what the solo icon should look like - perhaps the two should be paired visually in some way?
- When considering to-show-or-not-to-show for inheritance, I think my preference is definitely to show (but faded). But _not_ to show an icon when there is no attribute at all (so uncommenting your `if` but not your `else`). Whatever we do here, I think we should do for other attribute columns too, so we should factor that in to any decision making.
- I'm not convinced I like the "cross talk" between the solo and the mute column. What with inheritance, there's already quite a lot going on in the mute column, and I don't think showing the effects of solo in there too is a win. One concern with dropping that could be that it's harder to realise that mute is irrelevant because a solo exists (and possibly scrolled out of view). I think we could have a highlighted icon on the _header_ for solo in that case, so there's an always-visible alert that some soloing is happening.
- I wonder if we can go one step further with the double-click-or-enter-to-toggle? When creating an edit for the first time, could we enable the edit and set the state rather than showing the plug popup?
- Kindof separate, but I noticed a couple of things that bothered me about the LightEditor presentation generally :
    - The "Edited in this EditScope blue" doesn't stand out well enough when the cell is also selected. Should selection be fainter? Or a different colour entirely?
    - There's an annoying even-fainter-blue highlight on whatever Qt thinks the current cell is. We should get rid of this I think, or change the appearance so it can't be confused with selection or edit-scope-blue.

ericmehl commented on src/GafferScene/Light.cpp:
```diff
@@ -194,9 +195,13 @@ IECore::ConstCompoundObjectPtr Light::computeAttributes( const SceneNode::SceneP
+	if( mutePlug()->enabledPlug()->getValue() )
 	{
-		result->members()[g_lightMuteAttributeName] = g_true;
+		auto p = IECore::runTimeCast<const BoolPlug>( mutePlug()->valuePlug() );
```
Fixed in 842365bbe7824ac7e5cadd5c1d3740d9341a55d4

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -241,7 +241,10 @@ def __updateColumns( self ) :
-		self.__pathListing.setColumns( [ nameColumn ] + sectionColumns )
+		muteColumn = _GafferSceneUI._LightEditorMuteColumn(
+			GafferSceneUI.Private.AttributeInspector( self.__settingsNode["in"], self.__settingsNode["editScope"], "light:mute" )
+		)
```
Fixed in bc929bc4985c66a1467c3c68907453b384bf085d

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -232,6 +233,89 @@ class InspectorColumn : public PathColumn
+	public :
+		IE_CORE_DECLAREMEMBERPTR( MuteColumn )
+
+		MuteColumn( GafferSceneUI::Private::InspectorPtr inspector ) : InspectorColumn( inspector, "Mute" )
```
Fixed in 8e1e5f1d4573cc88c58219117d3ecce0a2966998

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -232,6 +233,89 @@ class InspectorColumn : public PathColumn
+
+					assert( plug );
+
+					plug->setValue( !plug->getValue() );
```
Fixed in 33a32c7edaba8bd7471a002fdd89e1c99317b08f

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -232,6 +233,89 @@ class InspectorColumn : public PathColumn
+
+					assert( plug );
+
+					plug->setValue( !plug->getValue() );
```
Fixed in 33a32c7edaba8bd7471a002fdd89e1c99317b08

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -232,6 +233,89 @@ class InspectorColumn : public PathColumn
+			else
+			{
+				result.icon = m_unMuteIconName;
+			}
```
Fixed in c76ba2c41986cb8ace24bcd01cf05e5cd68b22d3

ericmehl commented:
After some experimentation with icons and click behaviors, this is ready for another look and review. I fixed a test error in 6862e8cc1fc0dab0bdec1925e5ec69c8e4084ce9, and the changes to the icons and behavior is in c76ba2c41986cb8ace24bcd01cf05e5cd68b22d3 and c8c8a918774d7ab18ad46c93a859bb6545c41b7a.

Here's the current icon and editing scheme :
- No `light:mute` attribute : empty cell. Double clicking enables the attribute  and sets it to `True`.
- Inherited `light:mute` attribute : faded red icon for mute, faded low saturation green for un-mute. Double clicking enables the attribute and inverts the value that was inherited.
- `light:mute` attribute present : red icon for mute, low saturation green for un-mute. Double clicking inverts the value.

There's no indication of `solo`, that is going to be part of the solo column only.

There's a commit at the end that I don't particularly like, but I'm including it for others to consider. It adds highlighting to the mute icons when you hover over them. My main argument against it is that you can't see the button change color when you double click on it, so you can't tell if it registered until you move your mouse out of the cell. Hopefully users are accustomed to try double clicking it without a highlight cue, since the other cells are clickable and don't highlight either?

ericmehl commented:
My latest push has commit e1ea3782ac8caaa6de46c7fed75b343140cc4776 which tests the toggle behavior for the mute column. It's a little tedious to track the details of what's happening with each toggle, but hopefully it's pretty clear.

I think now we just need to settle on an icon - I'll play with that a bit more this afternoon.

ericmehl commented:
Alright, I have organized the commit madness and icon dogpile into one mostly nice string of commits.

For the icon, I did abandon the slash in favor of a vaguely studio-light / inspired-by-a-speaker-on-mute shape, while keeping the different color and opacity variations for different states. I think that might be easier to visually tie into the "solo" column when it arrives.

I updated the commit links in the resolved conversations above after the rebase. In addition to those there are commits :
- 09586758ff45bf1eb90071f6ab5c3147377311e9 : Adds the ability to toggle a mute state if it does not exist already. Prior to this commit, toggling a nonexistent `light:mute` attribute would give the plug popup
- 46d7600184d80fe82db8728b6c6272a4168f5cd1 : Adds icons to express inherited mute state
- 924481d1999c4fcfed3b9b64731883b19f72d662 : Adds highlighting icons and an undefined (blank) icon that is also used as a shim to prevent the top row resizing when it's mute is toggled
- e9985b23b396d41613b67fbd9a413bdc7b4d47d8 : Does the tricky bits for a toggle action on an undefined (missing) `light:mute` attribute. It looks for an ancestor that has a real opinion and sets it to the opposite of that.
- 98eaf13329368c278769833cb95d33bf9f2bbc8f : Adds a test for the toggle behaviors, and fixes a couple things to get them to pass.
- 6c837ee40d0b0e2a588b03c844951f8debdcf82b : Adds a popup box for the mute column if it can't be toggled. Now that a user can toggle mute state on a group and other non-light locations and expect an action in many cases, this seems helpful to guide them when their edit scope isn't set right.

a8df7d5cdba1bac6ba0b7332d6f3b1c02226c92c is a cherry-pick from #4994, which can drop out when that is merged.

ericmehl commented:
> [a8df7d5](https://github.com/GafferHQ/gaffer/commit/a8df7d5cdba1bac6ba0b7332d6f3b1c02226c92c) is a cherry-pick from #4994, which can drop out when that is merged.

Rebased and dropped that commit now that #4994 is merged, otherwise unchanged.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -359,6 +359,15 @@ def __editSelectedCells( self, pathListing, quickBoolean = True ) :
 						inspections.append( ( column.inspector(), inspection, parentInspection ) )
 
 		if len( inspections ) == 0 :
+			# \todo Add a way to identify columns that inspect attributes so this warning
```
I think we can already identify attribute-inspecting columns with `isinstance( column, InspectorColumn ) and isinstance( column.inspector(), AttributeInspector )`. But I don't understand what we're trying to do here - what am I missing?

ericmehl replied:
There's a situation where you won't be able to toggle because we couldn't find any inspectors. My intention is to give a popup message to that effect to but now I'm trying to recall where that situation comes up. I'll think more on it tonight.

ericmehl replied:
Right, the case I am catching is, for example, trying to toggle a mute on a group location with the Edit Scope set to `None`. Without this, it will do nothing and not give any feedback as to why. 

Since my original code comment is incorrect, and we can find out if the column is an attribute column, we could :
1. Give the popup warning for any attribute column
2. Give the same popup for any column, including shader parameters, even though those can never be set for non-lights
3. Give the popup for attributes, and a different worded one for parameters
4. Output a warning to the terminal instead
5. Remove this entirely

I favor 1, but am open to other opinions.

johnhaddon replied:
Do you prefer 1 because you don't see the need for 2 because nobody could reasonably expect that to _ever_ work, and they therefore don't need a warning? That sounds reasonable, but I'd like to play devil's advocate. We don't know what inspectors we might introduce in future, and eventually we intend to allow custom inspectors to be defined as extensions. And it might not be _completely_ obvious to a user why they shouldn't be able add a parameter on a group. So maybe the simpler version 2 where we warn for everything has some merit?

ericmehl replied:
> because you don't see the need for 2 because nobody could reasonably expect that to _ever_ work, and they therefore don't need a warning

Yeah, that's where I was coming from, it felt a bit noisy to me. But considering the plans for more columns and customization, the second option seems like a nice helping-hand. I can also imagine it handy for someone clicking / entering a row above or below what they really wanted, and a popup would help then too.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -440,6 +440,37 @@ def __toggleBoolean( self, inspections ) :
+		tweaks = []
+
+		with Gaffer.Context( self.getContext() ) as context :
+			for i in range( 0, len( columns ) ) :
```
We could ditch all the `[i]` indexing if we did this :

```
for selection, column in zip( pathListing.getSelection(), pathListing.getColumns() ) :
```

Might make the inner loop a tiny bit more readable?

ericmehl replied:
`zip` is new to me, I like it! Fixed in 3b1e785f90f9c76297491c2fa26e5e17f799f77e along with a few other spots.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -440,6 +440,37 @@ def __toggleBoolean( self, inspections ) :
+					context["scene:path"] = GafferScene.ScenePlug.stringToPath( path )
+					inspection = column.inspector().inspect()
+					if inspection is not None :
+						edit = inspection.acquireEdit()
```
Nooo! Calling `acquireEdit()` creates an edit, so merely by showing the popup menu with the "Remove Edits..." item in it, we are now _creating_ edits (albeit disabled)! Plus, `acquireEdit()` will throw if the thing isn't editable...

I _think_ we'll be OK if we first check `inspection.sourceType() == SourceType.EditScope` and `inspection.editable()`, but you might want to verify for yourself that there's no situation in which that will end up creating an edit.

ericmehl replied:
Fixed in 000eb3ff824f7e2c8f7adf3fecb7e3ac5df9ff35

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -440,6 +440,37 @@ def __toggleBoolean( self, inspections ) :
+
+	def __removeEdits( self, pathListing ) :
+
+		tweaks = self.__activeInspectionTweaks( pathListing )
```
This needs to be in an UndoScope.

ericmehl replied:
Fixed in 0e0764df41d4f84d50da9438bde416e50b02b7c9

johnhaddon commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -198,9 +198,20 @@ class InspectorColumn : public PathColumn
+
+			if( runTimeCast<const IECore::BoolData>( result.value ) )
+			{
+				toolTip += !toolTip.empty() ? "\nDouble-click to toggle" : "Double-click to toggle";
```
I think the tooltip would be a little more readable if we used `\n\n` to include a blank line between the "Source" part and the "Things you can do" part. That would match similar tooltips in the HierarchyView and NodeEditor.

ericmehl replied:
Fixed in 1b23e6defb06f657e1ea0fa85179ad9578cf3cc8

johnhaddon reviewed (changes requested):
Thanks Eric! Sorry for taking so long to look at this again - I've reviewed all the fixups and commented inline now. I also have some more icon quibbles, one which I feel confident about and one which will need consensus among the judges of the mute-solo-icon-pageant-2022 :

- On the faded "inherited value" icons, the "icon is hovered" border highlight shouldn't be faded - the border should be full strength. The value is inherited, but the hover is just as real as any other hover.
- I'm not loving the is-it-a-red-spotlight?-is-it-a-speaker? icon. I already think "mute" is a bit of weird metaphor for lights, and using a speaker-like icon pushes that a bit far for me. I would probably prefer the slash still, but am open to other ideas. Thinking to the next step, perhaps even a super-simple red-dot-for-mute, yellow-dot-for-solo might be sufficient? Anyway, I know UI bikeshedding is frustrating, so let's see if anyone agrees with me before changing anything.

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -251,13 +251,23 @@ class MuteColumn : public InspectorColumn
+				if( auto fullValue = fullAttributes->member<BoolData>( "light:mute" ) )
+				{
+					result.icon = fullValue->readable() ? m_muteFadedIconName : m_unMuteFadedIconName;
+				}
```
That sounds good - I have a note about adding a new `AttributeInspectorColumn`. Option 2 is done in 3ff4e995d41c8ac86c99f4401ca65d9d10a8f13a

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -267,12 +277,18 @@ class MuteColumn : public InspectorColumn
 
 	private :
 
+		const GafferScene::ScenePlugPtr m_scene;
```
Removed in 39847d8c81c3c1807081a542020a12c8bec68757

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -384,16 +404,30 @@ def __toggleBoolean( self, plugs, newValue ) :
+		soleValue = sole( values )
+		if soleValue is None and not all( v is None for v in values ) :
+			# No single value to use
+			return False
```
Fixed in 39847d8c81c3c1807081a542020a12c8bec68757 and 39847d8c81c3c1807081a542020a12c8bec68757

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -322,6 +322,8 @@ def __editSelectedCells( self, pathListing, quickBoolean = True ) :
 
 		columns = pathListing.getColumns()
 
+		# A list of tuples of GafferSceneUI.Inspector for the result, GafferSceneUI.Inspector.Result
```
Yeah, these were all a mess, I have a significant refactor in a273574c526a54cf6093a93fde345ddc6876b206 that I think is much clearer now.

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -334,27 +336,43 @@ def __editSelectedCells( self, pathListing, quickBoolean = True ) :
+			
 			performedEdit = False
 
 			if quickBoolean :
```
Fixed in 32e614c08da31c9647d2cd7205f8b5f0f4908be4

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -334,27 +336,43 @@ def __editSelectedCells( self, pathListing, quickBoolean = True ) :
 
 					if inspection is not None :
-						inspections.append( inspection )
+						parentInspection = None
```
Fixed as part of refactor in a273574c526a54cf6093a93fde345ddc6876b206

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -384,16 +404,30 @@ def __toggleBoolean( self, plugs, newValue ) :
+			) else None
+
+			newValue = not soleValue.value if soleValue is not None else not pParent
+			if isinstance( inspections[i][0], GafferSceneUI.Private.AttributeInspector ) :
```
Fixed as part of refactor in a273574c526a54cf6093a93fde345ddc6876b206

ericmehl commented on python/GafferSceneUITest/LightEditorTest.py:
```diff
@@ -0,0 +1,566 @@
+				}
+			),
+			"/groupUnMute" : (
+				imath.V3f(105, 110, 1),
```
Right, I had tried to stick with the public API, found that I wouldn't be able to go all the way, but forgot to come back and use the private `__editSelectedCells()`. Fixed in 743e2996e45764f9097f07b3412e5309cfa891c0

ericmehl commented on python/GafferSceneUITest/LightEditorTest.py:
```diff
@@ -0,0 +1,566 @@
+				GafferUI._qtAddress( widget._qtWidget().model() )
+			)
+			widget.setExpansion( IECore.PathMatcher( [ "/groupMute", "/groupUnMute" ] ) )
+			self.waitForIdle( 10000 )
```
Fixed in 743e2996e45764f9097f07b3412e5309cfa891c0 along with a general cleanup of UI manhandling.

johnhaddon commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -274,12 +274,24 @@ class MuteColumn : public InspectorColumn
 			{
-				auto fullAttributes = scenePath->getScene()->fullAttributes( scenePath->names() );
-				if( auto fullValue = fullAttributes->member<BoolData>( "light:mute" ) )
+				ScenePlug::PathScope pathScope( Context::current() );
```
This needs to be `scenePath->context()` rather than `Context::current()`. PathListingWidget doesn't know anything about contexts, so the current context is just going to be a default one.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/284acdd56ca77dc09f2ce3d06122cde6420b7656

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -408,27 +389,35 @@ def __toggleBoolean( self, inspections ) :
 
-		soleValue = sole( values )
+		valueData = [ i.value() for i in inspections ]
+		soleValue = sole( v.value if v is not None else False for v in valueData )
```
I haven't followed the logic through to see why, but this is still not matching the spreadsheet behaviour :

1. Select two cells, where one is muted and one is not.
2. Hit <kbd>Enter</kbd> to toggle. They each flip their current value, whereas the Spreadsheet flips them all to the same value.

ericmehl replied:
I added some tests for this case in https://github.com/GafferHQ/gaffer/pull/4948/commits/a502068ef81b71bf1572e9355efa9181f721de8f and fixed it in https://github.com/GafferHQ/gaffer/pull/4948/commits/0776038e2bc8e94d18050b0b588105fa264b5c9a

There are a couple other tests / fixes for another case that I came across today. Prior to this commit, if you toggled a light and its parent at the same time, the enabled state would not be correct. I was getting the parent value from pre-toggle state, which means it would be opposite of the post-toggle state if the parent and child are both toggled.

johnhaddon reviewed (commented):
Thanks for the fixups Eric. I've resolved most of the previous comments, but added a couple of new ones.

One other thing - how are you getting the "Fixed in \<xxx\>" links for your comments? I don't know if something has changed recently, but they now fire me off to looking at the commit in some generic GitHub commit page, unrelated to the PR. So I can't add PR comments there. I've found I need to right-click the SHA in the commit list and choose "Copy Link" to get a link that takes you to the relevant section of the "Files Changed" tab instead...

Cheers...
John

ericmehl commented:
> One other thing - how are you getting the "Fixed in <xxx>" links for your comments?

I copy them from the git helper in VS Code, which copies just the SHA and Github figures out the link when I paste it in. The problem might be that I rebased some later commits after leaving the comment. I checked that the SHA is the same, but maybe the link Github generated is stale after the rebase.

Or maybe I should copy from the Github web interface.

I'll be more careful in the future to not rebase in the middle of commenting and use the web links too.

ericmehl commented:
Strange, I'm still getting "this commit does not belong to any branch in this repository..." even though the same link in the commit list works fine.

ericmehl commented:
I have a new icon up, using a red and gray circle for mute / unmute. I also fixed the highlighting of the faded icons to be full strength.

Still to fix is the issue raised in #5019 which applies to mutes. And the case raised by @murraystevenson :
Add a light followed by two consecutive edit scopes. Turn mute on in the first via the Light Editor. Switch to the second edit scope and it won't be able to toggle off.

I have a fix in progress for that, but it needs some refinement still.

johnhaddon commented:
Here's another one along similar lines to Murray's problem, but simpler still :

</details>

```
import Gaffer
import GafferArnold
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 2, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["SpotLight"] = GafferArnold.ArnoldLight( "SpotLight" )
parent.addChild( __children["SpotLight"] )
__children["SpotLight"].loadShader( "spot_light" )
__children["SpotLight"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"] = Gaffer.EditScope( "EditScope" )
parent.addChild( __children["EditScope"] )
__children["EditScope"].addChild( Gaffer.BoxIn( "BoxIn" ) )
__children["EditScope"]["BoxIn"].setup( GafferScene.ScenePlug( "out", ) )
__children["EditScope"].addChild( GafferScene.ScenePlug( "in", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"].addChild( Gaffer.BoxOut( "BoxOut" ) )
__children["EditScope"]["BoxOut"].setup( GafferScene.ScenePlug( "in", ) )
__children["EditScope"].addChild( GafferScene.ScenePlug( "out", direction = Gaffer.Plug.Direction.Out, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"].addChild( Gaffer.BoolPlug( "enabled", defaultValue = True, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"].addChild( GafferScene.SceneProcessor( "AttributeEdits" ) )
__children["EditScope"]["AttributeEdits"].addChild( Gaffer.Spreadsheet( "Spreadsheet" ) )
__children["EditScope"]["AttributeEdits"]["Spreadsheet"]["rows"].addColumn( Gaffer.TweakPlug( Gaffer.BoolPlug( "value", defaultValue = True, ), "light_mute", ), adoptEnabledPlug = True )
__children["EditScope"]["AttributeEdits"]["Spreadsheet"]["rows"].addRows( 1 )
__children["EditScope"]["AttributeEdits"].addChild( GafferScene.PathFilter( "PathFilter" ) )
__children["EditScope"]["AttributeEdits"].addChild( GafferScene.AttributeTweaks( "AttributeTweaks" ) )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["tweaks"].addChild( Gaffer.TweakPlug( Gaffer.BoolPlug( "value", defaultValue = True, ), "tweak", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["EditScope"]["AttributeEdits"].addChild( Gaffer.Spreadsheet.RowsPlug( "edits", ) )
__children["EditScope"]["AttributeEdits"]["edits"].addColumn( Gaffer.TweakPlug( Gaffer.BoolPlug( "value", defaultValue = True, ), "light_mute", ), adoptEnabledPlug = True )
__children["EditScope"]["AttributeEdits"]["edits"].addRows( 1 )
__children["SpotLight"]["name"].setValue( 'spotLight' )
__children["SpotLight"]["mute"]["value"].setValue( True )
__children["SpotLight"]["mute"]["enabled"].setValue( True )
__children["SpotLight"]["__uiPosition"].setValue( imath.V2f( -1.7603898, 14.1000004 ) )
__children["EditScope"]["BoxIn"]["__in"].setInput( __children["EditScope"]["in"] )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxIn"]["__in"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxIn"]["__in"], 'description', 'The output scene.' )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxIn"]["__in"], 'nodule:type', 'GafferUI::StandardNodule' )
__children["EditScope"]["in"].setInput( __children["SpotLight"]["out"] )
Gaffer.Metadata.registerValue( __children["EditScope"]["in"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["EditScope"]["in"], 'description', 'The output scene.' )
Gaffer.Metadata.registerValue( __children["EditScope"]["in"], 'nodule:type', 'GafferUI::StandardNodule' )
__children["EditScope"]["BoxOut"]["in"].setInput( __children["EditScope"]["AttributeEdits"]["out"] )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxOut"]["__out"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxOut"]["__out"], 'description', 'The output scene.' )
Gaffer.Metadata.registerValue( __children["EditScope"]["BoxOut"]["__out"], 'nodule:type', 'GafferUI::StandardNodule' )
__children["EditScope"]["BoxOut"]["passThrough"].setInput( __children["EditScope"]["BoxIn"]["out"] )
__children["EditScope"]["BoxOut"]["enabled"].setInput( __children["EditScope"]["enabled"] )
__children["EditScope"]["out"].setInput( __children["EditScope"]["BoxOut"]["__out"] )
Gaffer.Metadata.registerValue( __children["EditScope"]["out"], 'nodule:color', imath.Color3f( 0.240099996, 0.339399993, 0.485000014 ) )
Gaffer.Metadata.registerValue( __children["EditScope"]["out"], 'description', 'The output scene.' )
Gaffer.Metadata.registerValue( __children["EditScope"]["out"], 'nodule:type', 'GafferUI::StandardNodule' )
__children["EditScope"]["__uiPosition"].setValue( imath.V2f( -0.261228979, 5.76796913 ) )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"], 'editScope:processorType', 'AttributeEdits' )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"], 'icon', 'editScopeProcessorNode.png' )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"], 'nodeGadget:color', imath.Color3f( 0.187600002, 0.390799999, 0.600000024 ) )
__children["EditScope"]["AttributeEdits"]["out"].setInput( __children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["out"] )
__children["EditScope"]["AttributeEdits"]["in"].setInput( __children["EditScope"]["BoxIn"]["out"] )
__children["EditScope"]["AttributeEdits"]["Spreadsheet"]["selector"].setValue( '${scene:path}' )
__children["EditScope"]["AttributeEdits"]["Spreadsheet"]["rows"].setInput( __children["EditScope"]["AttributeEdits"]["edits"] )
__children["EditScope"]["AttributeEdits"]["PathFilter"]["paths"].setInput( __children["EditScope"]["AttributeEdits"]["Spreadsheet"]["enabledRowNames"] )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["in"].setInput( __children["EditScope"]["AttributeEdits"]["in"] )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["filter"].setInput( __children["EditScope"]["AttributeEdits"]["PathFilter"]["out"] )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["localise"].setValue( True )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["ignoreMissing"].setValue( True )
__children["EditScope"]["AttributeEdits"]["AttributeTweaks"]["tweaks"]["tweak"].setInput( __children["EditScope"]["AttributeEdits"]["Spreadsheet"]["out"]["light_mute"] )
__children["EditScope"]["AttributeEdits"]["edits"][0]["cells"]["light_mute"]["value"]["name"].setValue( 'light:mute' )
__children["EditScope"]["AttributeEdits"]["edits"][0]["cells"]["light_mute"]["value"]["enabled"].setValue( False )
__children["EditScope"]["AttributeEdits"]["edits"][0]["cells"]["light_mute"]["value"]["mode"].setValue( 5 )
__children["EditScope"]["AttributeEdits"]["edits"][1]["name"].setValue( '/spotLight' )
__children["EditScope"]["AttributeEdits"]["edits"][1]["cells"]["light_mute"]["value"]["name"].setValue( 'light:mute' )
__children["EditScope"]["AttributeEdits"]["edits"][1]["cells"]["light_mute"]["value"]["enabled"].setValue( False )
__children["EditScope"]["AttributeEdits"]["edits"][1]["cells"]["light_mute"]["value"]["mode"].setValue( 5 )
__children["EditScope"]["AttributeEdits"]["edits"][1]["cells"]["light_mute"]["value"]["value"].setValue( False )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"]["edits"], 'spreadsheet:defaultRowVisible', False )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"]["edits"][0], 'spreadsheet:rowNameWidth', 300 )
Gaffer.Metadata.registerValue( __children["EditScope"]["AttributeEdits"]["edits"][0]["cells"]["light_mute"]["value"], 'tweakPlugValueWidget:allowCreate', True )


del __children

```

</details>

Focus the EditScope and set it as the scope in the LightEditor, and it's not possible to unmute the light.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -392,35 +393,55 @@ def __toggleBoolean( self, inspectors, inspections ) :
+
+				if isinstance( inspector, GafferSceneUI.Private.AttributeInspector ) :
+					fullAttributes = self.__settingsNode["in"].fullAttributes( path[:-1] )
+					parentValueData = fullAttributes.get( "light:mute", None )
```
We can't hardcode "light:mute" here, because this code is intended to run for any boolean attribute (or parameter).

johnhaddon replied:
This is still hardcoded to `light:mute`.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -392,35 +393,55 @@ def __toggleBoolean( self, inspectors, inspections ) :
-						enabled = True
+				if isinstance( inspector, GafferSceneUI.Private.AttributeInspector ) :
+					fullAttributes = self.__settingsNode["in"].fullAttributes( path[:-1] )
+					parentValueData = fullAttributes.get( "light:mute", None )
```
Again, we can't hardcode anything here.

johnhaddon reviewed (changes requested):
Thanks for the update Eric, new icons LGTM. I've resolved my previous comments but added two more inline - should be simple ones though I think.

johnhaddon commented:
Just had a conversation with @ericmehl in which I might have successfully argued us into a much simpler version of the toggle interaction. I'm going to try to repeat the argument here for posterity, and so @murraystevenson can point out the flaws when he's back after Christmas.

First, a recap. We had all agreed with the idea of automatically eliding redundant attributes, to keep scenes clean and to provide maximum opportunity of creating edits high in the hierarchy, without them being overridden by descendants. Eric has implemented the logic for this already, something along the lines of this pseudocode :

```
newValue = not attributeValue( thisLocation ) # toggle
inheritedValue = attributeValue( parentLocation )
if newValue != inheritedValue :
    makeAttributeEdit( thisLocation, newValue )
else :
    # We don't want a local attribute, because we can inherit the value we want
    # from our parent. Disable any existing edit at this location.
    existingEdit = acquireAttributeEdit( thisLocation, /* createIfMissing = */ False )
    if existingEdit : existingEdit["enabled"].setValue( False )
```

This logic has a flaw, exposed by Murrays "two EditScopes in a row" and my "EditScope after Light" examples above. The problem is simple : if the attribute exists already in the input to the EditScope, disabling the edit will mean we keep that input value. To inherit the parent value cleanly, we need to _remove_ the local attribute forcefully, with a Tweak in Remove mode. This would be easy enough to do, but it has made me rethink this whole enterprise.

With the fix above, a double-click-to-toggle-mute operation could now lead to any of three different states in the EditScope :

A. An enabled edit, with the new value.
B. No edit.
C. An edit to Remove the value.

I think there are a few potential concerns here :

1. The state we choose is dependent on the input state. For B and C, if the input to the EditScope changes, that state may no longer reproduce the result the user thinks they asked for. This seems to have a lot of potential for confusion, given that the whole point of Gaffer is to run the same graph on different inputs, in different contexts. "Hey, I muted that light in my shot edits!"..."Sure sure, but at the time you muted it, it happened to be muted already in the upstream light rig, so Gaffer didn't bother".
2. The expected semantics of double-click are "Edit". It would seem surprising that it would instead _remove_ an existing edit. The rules above seem too complex to have to explain to users for such a seemingly simple operation as double-click.

I think we've partly ended up here because `light:mute` was the first attribute we considered, and our expectation was that generally it should be authored sparingly, and that you'd only really want to author `True` values. But that doesn't extend to other boolean attributes, and this double-click-to-toggle feature applies to all of them.

So I'm advocating for a much dumber version, where double-click just does A, authoring an edit which stomps the new value unconditionally. Now, I _do_ concede that this might feel a bit rubbish when interacting with the mute column. But I think it's the only approach which is explainable and predictable enough to work as a general purpose feature. And I think we can win back some of the cleanliness of attribute state we're hoping for through other features :

i. A `D` hotkey that disables edits in the selected cells, attached to the new menu item that Eric added already. I think I could quickly get into the habit of "select cells, hit D" to remove mute states. I propose we add this now, in this PR.
ii. Exposing the `Remove` mode for attribute edits, so folks can easily clean up incoming attributes they don't want. And connecting it to the `Delete` hotkey for ease of use. I don't think this is super-important for this use case, but it's easy enough to add now, or do soon.
iii. A "Remove Redundant Edits" menu item, that removes edits that are unnecessary through inheritance (from ancestors in the scene) or input (from upstream nodes). This would allow folks to clean things up, but crucially would make it obvious what is happening rather than being hidden behind an innocent double-click. I'd call this out of scope for this PR, but hopefully it demonstrates that we're not abandoning the idea of clean scenes. We're just wanting to make the semantics of each operation much clearer.

ericmehl commented:
> A `D` hotkey that disables edits in the selected cells, attached to the new menu item that Eric added already

I added the `D` hotkey, and after trying it I immediately was inclined to press it again to get the edit back. That's probably influenced by the spreadsheet behavior, which we are matching in many other ways in the Light Editor. I wonder if "Remove Edit" should become an enabled toggle as it is in the spreadsheet?

The main difference I guess is that unlike a spreadsheet, not all Light Editor cells can be toggled on - there may be no edit to toggle. And the disabled state is different from a spreadsheet, so it's not immediately obvious what can and cannot be toggled on.

I'll add the `Delete` behavior and then push those changes up so anyone interested can try it out.

ericmehl commented:
My latest push has the new behavior @johnhaddon described above. Toggling an attribute always results in definite toggle, no disabling of tweaks unless done by the user.

I renamed the "Remove Edit" to "Disable Edit". I think that's an improvement because it's more precisely what is happening, and it corresponds better to the `D` hotkey associated with the action. It also differentiates is from the other new item, "Remove Attribute".

I currently allow disabling a `Remove` attribute. I'm ambivalent as to whether or not that should be allowed or disallowed.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +370,51 @@ def __editSelectedCells( self, pathListing ) :
+					plug["value"].setValue( newValue )
+					plug["enabled"].setValue( True )
+				else :
+					plug.setValue( newValue )
```
This needs to go in an UndoScope.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/554e1f5e381afa936daf87228cb0e9e800bf2ed2

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -415,6 +415,37 @@ def __toggleBoolean( self, inspectors, inspections ) :
+
+	def __removeEdits( self, pathListing ) :
+
+		with Gaffer.UndoScope( self.__settingsNode["in"].getInput().ancestor( Gaffer.ScriptNode ) ) :
```
We have simpler access to the ScriptNode via `self.scriptNode()`.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/3494800095112a1673280dcfe1149485b5a5177c

johnhaddon commented on Changes.md:
```diff
@@ -21,7 +21,7 @@ Features
 - Light Editor
   - Added a column for toggling a light's mute attribute on and off.
-  - Added a "Remove Edit" command to the right-click menu to easily disable tweaks in Edit Scopes.
+  - Added a "Delete Edit" command to the right-click menu, shortcut <kbd>D</kbd>, to easily disable tweaks in Edit Scopes.
```
Delete -> Disable

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/f5e07ea1421f5949ce06f2b0021b73365e5c3428

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -309,10 +309,17 @@ def __buttonDoubleClick( self, pathListing, event ) :
+				return True
+
+			if event.key == "D" :
+				self.__removeEdits( pathListing )
```
Would be good to rename `__removeEdits` to `__disableEdits` so we're using the same terminology throughout.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/1e84f34dc950b00a745d9790dc2f39fd6df6ac31 and https://github.com/GafferHQ/gaffer/pull/4948/commits/3494800095112a1673280dcfe1149485b5a5177c

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -444,6 +451,32 @@ def __activeInspectionTweaks( self, pathListing ) :
 
 		return tweaks
 
+	def __removableAttributeTweaks( self, pathListing ) :
+
+		tweaks = []
+
+		with Gaffer.Context( self.getContext() ) as context :
+			for columnSelection, column in zip( pathListing.getSelection(), pathListing.getColumns() ) :
+				if not isinstance( column, _GafferSceneUI._LightEditorInspectorColumn ) :
+					continue
+				for path in columnSelection.paths() :
+					context["scene:path"] = GafferScene.ScenePlug.stringToPath( path )
+					inspection = column.inspector().inspect()
+					if (
+						inspection is not None and
+						inspection.editable() and
+						inspection.sourceType() == GafferSceneUI.Private.Inspector.Result.SourceType.EditScope
+					) :
+						edit = inspection.acquireEdit()
+						if (
+							isinstance( column.inspector(), GafferSceneUI.Private.AttributeInspector ) and
+							isinstance( edit, Gaffer.TweakPlug ) and
+							edit["mode"].getValue() != Gaffer.TweakPlug.Mode.Remove
+						) :
+							tweaks.append( edit )
+
+		return tweaks
```
This seems to mean that we can only delete attributes in an EditScope if they have already been edited in that scope. I'd like to be able to delete _any_ attribute in an EditScope, by authoring a _new_ tweak as necessary. Is there any reason we can't do that?

ericmehl replied:
Yes, that's definitely a good change to make. Removing the test of `inspection.sourceType() == GafferSceneUI.Private.Inspector.Result.SourceType.EditScope` allows the edit scope to delete any attribute.

It also allows you to delete (set the tweak mode to `Remove` to be more precise) an attribute that doesn't exist. I'm not sure if that's desirable or not. I think perhaps it is, with the use-case being "I'm working on a shot setup and want to make sure this attribute is removed, even if someone upstream of my shot adds it in the future." Do you think that's worth keeping?

ericmehl replied:
I reworked `__removableAttributeTweaks()` in https://github.com/GafferHQ/gaffer/pull/4948/commits/31a45171fb615016ea10ba86e2e8e2485154a14f and I think it's looking much better now. And by returning an `inspector` list instead of plugs, we can avoid creating edits until they are really needed, not just from checking to see if the menu item is enabled.

johnhaddon replied:
Yep - that seems like a nice way of approaching it. I think we need to do the same "is the whole selection removable?" check in the <kbd>Delete</kbd> shortcut as well though. Perhaps it would be even better if `__removableAttributeInspections` simply returned an empty list if it found anything that wasn't removable? Then the check is centralised and the mistake can't be repeated elsewhere...

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -549,6 +590,14 @@ def __buttonPress( self, pathListing, event ) :
 				}
 			)
+			menuDefinition.append(
+				"Remove Attribute",
```
We seem to have inconsistent terminology between the DeleteAttributes node and the AttributeTweaks node and its  Remove mode. I think my preference would be to standardise on "Delete", but I can't really articulate why. Any thoughts?

ericmehl replied:
Just posting here that we decided offline to not worry about changing this, at least partly because it might chain react into changing the `Remove` tweak mode enum.

johnhaddon reviewed (changes requested):
Sorry for the tardy review Eric. I had intended to get this one done before you started back...

ericmehl commented:
Along with fixups from the comments above, I took broader look at the `Remove Attributes` and `Disable Edits` code and I think those are now enabled in the right circumstances, and not creating edits where they aren't needed.

Ready for another look!

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -436,6 +578,28 @@ def __buttonPress( self, pathListing, event ) :
+				{
+					"command" : functools.partial( self.__removeAttributes, pathListing ),
+					"active" : lambda : len( self.__removableAttributeInspections( pathListing ) ) == sum( [ p.size() for p in pathListing.getSelection() ] ),
+					"shortCut" : "Delete",
```
Looking at similar menu items, I think this needs to be `"Backspace, Delete"`, and `__keyPress` needs to be updated to match that.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/27a4928eaa58773ca5832d4ae05aac2dd0d4a666

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +382,120 @@ def __editSelectedCells( self, pathListing ) :
+					not isinstance( column, _GafferSceneUI._LightEditorInspectorColumn ) or
+					not isinstance( column.inspector(), GafferSceneUI.Private.AttributeInspector )
+				) :
+					continue
```
If we return `[]` here, then we can remove the later check that the list is the same length as the selection.

ericmehl replied:
We still need to iterate over all of the columns to see if they have a selection, so I broke this conditional into two parts in https://github.com/GafferHQ/gaffer/pull/4948/commits/8e2b59b2402e01a4ee35db5e58913b67bf87ceb9. That commit also adds the early out and simplified testing in the `active` lambda.

johnhaddon replied:
I think there's still a minor bug in this, because the menu items remain active when there is an additional selection in the "Name" column. I'm going to merge anyway though, as better to include what we have in the upcoming release than hold it up any more.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +382,120 @@ def __editSelectedCells( self, pathListing ) :
+							( isinstance( source, Gaffer.TweakPlug ) and source["mode"].getValue() != Gaffer.TweakPlug.Mode.Remove ) or
+							editScope is not None
+						) :
+							inspections.append( inspection )
```
For the above to make sense, we'd need an `else : return []` here too.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/8e2b59b2402e01a4ee35db5e58913b67bf87ceb9

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +382,120 @@ def __editSelectedCells( self, pathListing ) :
+
+		return True
+
+	def __activeInspectionTweaks( self, pathListing ) :
+
+		tweaks = []
+
+		with Gaffer.Context( self.getContext() ) as context :
+			for columnSelection, column in zip( pathListing.getSelection(), pathListing.getColumns() ) :
+				if not isinstance( column, _GafferSceneUI._LightEditorInspectorColumn ) :
+					continue
+				for path in columnSelection.paths() :
+					context["scene:path"] = GafferScene.ScenePlug.stringToPath( path )
+					inspection = column.inspector().inspect()
+					if inspection is not None and inspection.editable() :
+						source = inspection.source()
+						editScope = self.__settingsNode["editScope"].getInput()
+						if (
+							isinstance( source, ( Gaffer.TweakPlug, Gaffer.NameValuePlug ) ) and
+							( editScope is None or editScope.node().isAncestorOf( source ) ) and
+							source["enabled"].getValue()
+						) :
+							tweaks.append( source )
+
+		return tweaks
```
We could use the same early `return []` that I suggested in `__removableAttributeInspections()` here too.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/fa15387c8165b2b9477b11732369f4d779d3ed10

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -302,48 +309,69 @@ def __buttonDoubleClick( self, pathListing, event ) :
 
-		columns = pathListing.getColumns()
+			if event.key == "Delete" :
+				self.__removeAttributes( pathListing )
```
This needs the same "is the whole selection deletable?" check that we have in the menu building code. But I've made a suggestion in `__removableAttributeInspections` that would make this moot. The same applies to the disable shortcut...

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/4948/commits/182b1e8f711fbdaf4e3096a88b86bd555810344f and https://github.com/GafferHQ/gaffer/pull/4948/commits/00da3c790ea52e7635bb5cfa13fdb92cf611012e

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +382,120 @@ def __editSelectedCells( self, pathListing ) :
+			for columnSelection, column in zip( pathListing.getSelection(), pathListing.getColumns() ) :
+				if (
+					not isinstance( column, _GafferSceneUI._LightEditorInspectorColumn ) or
+					not isinstance( column.inspector(), GafferSceneUI.Private.AttributeInspector )
```
It baffled me that this check was sufficient to stop people trying to delete shader parameters, because ParameterInspector derives form AttributeInspector. It turns out that the reason it works is because there's a bug in the bindings, and we've bound ParameterInspector as a subclass of Inspector and not AttributeInspector.

ericmehl replied:
I think I solved half of this in https://github.com/GafferHQ/gaffer/pull/4948/commits/66b80ab6f007e978327ebe4f3712ad9a1a9f8a68 by comparing directly to the `type()` of column.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -354,6 +382,120 @@ def __editSelectedCells( self, pathListing ) :
+			for inspection in inspections :
+				tweak = inspection.acquireEdit()
+				tweak["enabled"].setValue( True )
+				tweak["mode"].setValue( Gaffer.TweakPlug.Mode.Remove )
```
Once we've removed an attribute, viewing the attribute history gives this warning : ```PathListingWidget : Could not find file "removeSmall.png"```.

ericmehl replied:
Added the dotted-line box to indicate removal in https://github.com/GafferHQ/gaffer/pull/4948/commits/a28d7e853342e0260c8afa2f90b3be41392b4fb2

johnhaddon reviewed (changes requested):
Thanks Eric! I spotted a few more things (see inline comments) but I'm not sure we need to address them all in this PR - some could be done as followups. I'd also like to get a new set of releases out tomorrow, and it would be good to include this in `1.2.0.0a2`, so let's have a chat about what we do in this PR and what we defer...

ericmehl commented:
I have the latest round of fixes in place and noted above. Hopefully we're passing and with another review I can squash down!

johnhaddon commented:
Thanks Eric! I tried a `git rebase -i --autosquash` but it resulted in loads of conflicts, so I've done a manual rebase to get it all squashed down. To be safe I did a diff against the original and I think all is well. Also moved all the Changes.md additions into the new version. And I've added [one commit](https://github.com/GafferHQ/gaffer/pull/4948/commits/e6f4572db423490f52551b99bf9e87fa6ca314c4) of my own that gets rid of a circular reference - hopefully all makes sense. Right, I'm gonna merge now and get this into a release!
